exchanges/coinbase/coinbase_test.go

```diff
@@ -0,0 +1,2099 @@
+	assert.ErrorIs(t, err, order.ErrUnsupportedStatusType)
+}
+
+func TestStringToStandardType(t *testing.T) {
+	t.Parallel()
+	resp, _ := stringToStandardType("LIMIT_ORDER_TYPE")
+	assert.Equal(t, order.Limit, resp)
+	resp, _ = stringToStandardType("MARKET_ORDER_TYPE")
+	assert.Equal(t, order.Market, resp)
+	resp, _ = stringToStandardType("STOP_LIMIT_ORDER_TYPE")
+	assert.Equal(t, order.StopLimit, resp)
+	_, err := stringToStandardType("")
+	assert.ErrorIs(t, err, order.ErrUnrecognisedOrderType)
+}
+
+func TestStringToStandardAsset(t *testing.T) {
+	t.Parallel()
+	resp, _ := stringToStandardAsset("SPOT")
+	assert.Equal(t, asset.Spot, resp)
+	resp, _ = stringToStandardAsset("FUTURE")
+	assert.Equal(t, asset.Futures, resp)
+	_, err := stringToStandardAsset("")
+	assert.ErrorIs(t, err, asset.ErrNotSupported)
+}
+
+func TestStrategyDecoder(t *testing.T) {
+	t.Parallel()
+	resp, _ := strategyDecoder("IMMEDIATE_OR_CANCEL")
+	assert.True(t, resp.Is(order.ImmediateOrCancel))
+	resp, _ = strategyDecoder("FILL_OR_KILL")
+	assert.True(t, resp.Is(order.FillOrKill))
+	resp, _ = strategyDecoder("GOOD_UNTIL_CANCELLED")
+	assert.True(t, resp.Is(order.GoodTillCancel))
+	resp, _ = strategyDecoder("GOOD_UNTIL_DATE_TIME")
+	assert.True(t, resp.Is(order.GoodTillDay|order.GoodTillTime))
+	_, err := strategyDecoder("")
+	assert.ErrorIs(t, err, errUnrecognisedStrategyType)
+}
+
+func TestBase64URLEncode(t *testing.T) {
+	t.Parallel()
+	resp := base64URLEncode([]byte{byte(252), byte(253), byte(254), byte(255)})
+	assert.Equal(t, "_P3-_w", resp)
+}
+
+func TestProcessFundingData(t *testing.T) {
+	t.Parallel()
+	accHistory := []DeposWithdrData{
+		{
+			Type: "unknown",
+		},
+		{
+			Type: "TRANSFER_TYPE_WITHDRAWAL",
+		},
+	}
+	cryptoHistory := []TransactionData{
+		{
+			Type: "receive",
+		},
+		{
+			Type: "send",
+		},
+	}
+	_, err := e.processFundingData(accHistory, cryptoHistory)
+	assert.ErrorIs(t, err, errUnknownTransferType)
+	accHistory[0].Type = "TRANSFER_TYPE_DEPOSIT"
+	resp, err := e.processFundingData(accHistory, cryptoHistory)
+	require.NoError(t, err)
+	assert.NotEmpty(t, resp)
+}
+
+func TestChannelName(t *testing.T) {
+	_, err := channelName(&subscription.Subscription{})
+	assert.ErrorIs(t, err, subscription.ErrNotSupported)
+	_, err = channelName(&subscription.Subscription{Channel: subscription.HeartbeatChannel})
+	assert.NoError(t, err)
+}
+
+func exchangeBaseHelper(e *Exchange) error {
+	cfg := config.GetConfig()
+	err := cfg.LoadConfig("../../testdata/configtest.json", true)
+	if err != nil {
+		return err
+	}
+	gdxConfig, err := cfg.GetExchangeConfig("Coinbase")
+	if err != nil {
+		return err
+	}
+	if apiKey != "" {
+		gdxConfig.API.Credentials.Key = apiKey
+		gdxConfig.API.Credentials.Secret = apiSecret
+		gdxConfig.API.AuthenticatedSupport = true
+		gdxConfig.API.AuthenticatedWebsocketSupport = true
+	}
+	e.Websocket = sharedtestvalues.NewTestWebsocket()
+	err = e.Setup(gdxConfig)
+	if err != nil {
+		return err
+	}
+	return nil
+}
```

Comment:
Can be refactored to

```suggestion
func exchangeBaseHelper(e *Exchange) error {
	if err := testexch.Setup(e); err != nil {
		return err
	}

	if apiKey != "" {
		e.SetCredentials(apiKey, apiSecret, "", "", "", "")
		e.API.AuthenticatedSupport = true
		e.API.AuthenticatedWebsocketSupport = true

	}
	return nil
}
```

Then you can rid a lot of the `SetDefault` calls since `testexch.Setup` will do it for you